certificates/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Utilities for validating X509 certificates, used by both nexus and wicketd.
+
+use omicron_common::api::external::Error;
+use openssl::asn1::Asn1Time;
+use openssl::pkey::PKey;
+use openssl::x509::X509;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CertificateError {
+    #[error("Failed to parse certificate: {0}")]
+    BadCertificate(openssl::error::ErrorStack),
+
+    #[error("Certificate exists, but is empty")]
+    CertificateEmpty,
+
+    #[error("Certificate exists, but is expired")]
+    CertificateExpired,
+
+    #[error("Failed to parse private key")]
+    BadPrivateKey(openssl::error::ErrorStack),
+
+    #[error("Certificate and private key do not match")]
+    Mismatch,
+
+    #[error("Unexpected error: {0}")]
+    Unexpected(openssl::error::ErrorStack),
+}
+
+impl From<CertificateError> for Error {
+    fn from(error: CertificateError) -> Self {
+        use CertificateError::*;
+        match error {
+            BadCertificate(_) | CertificateEmpty | CertificateExpired
+            | Mismatch => Error::InvalidValue {
+                label: String::from("certificate"),
+                message: error.to_string(),
+            },
+            BadPrivateKey(_) => Error::InvalidValue {
+                label: String::from("private-key"),
+                message: error.to_string(),
+            },
+            Unexpected(_) => {
+                Error::InternalError { internal_message: error.to_string() }
+            }
+        }
+    }
+}
+
+pub struct CertificateValidator {
+    validate_expiration: bool,
+}
+
+impl Default for CertificateValidator {
+    fn default() -> Self {
+        Self { validate_expiration: true }
+    }
+}
+
+impl CertificateValidator {
+    /// Disable validation of certificate expiration dates.
+    ///
+    /// This exists to support basic certificate validation even before time is
+    /// available (e.g., before the rack has been initialized and NTP has
+    /// started).
+    pub fn danger_disable_expiration_validation(&mut self) {
+        self.validate_expiration = false;
+    }
+
+    /// Validate that we can parse the cert chain, that the key matches, and
+    /// that the certs in the chain are not expired (unless we have disabled
+    /// expiration validation).
+    //
+    // TODO-completeness: Can we take an optional hostname and validate that
+    // this cert will work for it? (This might be a paramter to this method, or
+    // could be an additional property on `Self`.)
+    pub fn validate(
+        &self,
+        certs: &[u8],
+        key: &[u8],
```

Comment:
I think it'd be worth a note in the doc comment about what format these are supposed to be in.

Reply:
Added in ec5db70